build/resolve_buck.py

```diff
@@ -0,0 +1,135 @@
+BUCK_VERSION_DATE = "2024-02-15"
+BUCK_VERSION_LINUX_GNU_X86 = "38f7c508bf1b87bcdc816bf56d1b9f2d2411c6be"
+BUCK_VERSION_DARWIN_ARM64 = "99e407b49dc432eda0cbddd67ea78346"
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        description="Locates or downloads the appropriate version of buck2.",
+    )
+    parser.add_argument(
+        "--buck2",
+        default="",
+        help="Explicit 'buck2' path to use.",
+    )
+    parser.add_argument(
+        "--cache_dir",
+        help="Directory to cache downloaded versions of buck2.",
+    )
+    return parser.parse_args()
+
+
+def resolve_buck2(args: argparse.Namespace) -> str:
+    # Find buck2, in order of priority:
+    #  1) Explicit buck2 argument
+    #  2) System buck2 (if correct version)
+    #  3) Download buck2
+
+    # Determine the target buck2 version string according to the current
+    # platform. If the platform isn't linux or darwin, we won't perform
+    # any version validation.
+    target_buck_version = ""
+    buck_archive_name = ""
+
+    machine = platform.machine().lower()
+    is_x86 = machine == "x86" or machine == "x86_64" or machine == "amd64"
+    is_arm64 = machine == "arm64"
+
+    if sys.platform.startswith("linux") and is_x86:
+        buck_archive_name = "buck2-x86_64-unknown-linux-gnu.zst"
+        target_buck_version = BUCK_VERSION_LINUX_GNU_X86
+    elif sys.platform.startswith("darwin") and is_arm64:
+        buck_archive_name = "buck2-aarch64-apple-darwin.zst"
+        target_buck_version = BUCK_VERSION_DARWIN_ARM64
+    else:
+        return args.buck2 or "buck2"
+
+    if args.buck2:
+        # If we have an explicit buck2 arg, check the version and warn if
+        # there's a mismatch.
+        ver = buck_util.get_buck2_version(args.buck2)
+        if ver != target_buck_version:
+            print(
+                f'The provided buck2 binary reports version "{ver}", but '
+                f'ExecuTorch needs version "{target_buck_version}". Ensure '
+                "that the correct buck2 version is installed or avoid "
+                "explicitly passing the BUCK2 version to automatically "
+                "download the correct version.",
+                file=sys.stderr,
+            )
+        return args.buck2
+    else:
+        # Look for system buck2 and check version. Note that this can return
+        # None.
+        ver = buck_util.get_buck2_version("buck2")
+        if ver == target_buck_version:
+            # Use system buck2.
+            return "buck2"
+        else:
+            # Download buck2 or used previously cached download.
+
+            cache_dir = Path(args.cache_dir)
+            buck2_local_path = (
+                (cache_dir / f"buck2-{target_buck_version}").absolute().as_posix()
+            )
+            buck2_local_zstd_path = (
+                (cache_dir / f"{buck2_local_path}.zstd").absolute().as_posix()
+            )
+
+            if os.path.isfile(buck2_local_path):
```

Comment:
This could be a previously-downloaded copy of an old version of buck2, if we just bumped the version and the user recently synced the repo.

Reply:
I am naming the files with the version hash, so it should be okay.